### PR TITLE
support x509 certificate

### PR DESCRIPTION
**Pull request by ZhijunZhao**
@zikalino @yaohaizh @yaweiw @xscript @aovechki

### CODE REVIEW

File: Lesson3/package.json [hunk 1]
```diff
@@ -9,7 +9,7 @@
     "azure-storage": "^1.3.0",
     "get-gulp-args": "0.0.1",
     "gulp": "^3.9.1",
-    "gulp-common": "Azure/gulp-common.git#v0.10.9",
+    "gulp-common": "Azure/gulp-common.git#zhijzhao-x509",
```
Comment (xscript) [inline Lesson3/package.json]: ?

Comment (ZhijunZhao) [inline Lesson3/package.json]: just temporary

File: Lesson3/app/main.c [hunk 2]
```diff
@@ -48,7 +50,7 @@ static void sendCallback(IOTHUB_CLIENT_CONFIRMATION_RESULT result, void *userCon
     messagePending = false;
 }
 
-static void sendMessageAndBlink(IOTHUB_CLIENT_LL_HANDLE iotHubClientHandle, char* deviceId)
+static void sendMessageAndBlink(IOTHUB_CLIENT_LL_HANDLE iotHubClientHandle, char *deviceId)
```
Comment (xscript) [inline Lesson3/app/main.c]: Function name convention for C functions should be like `function_name_style`, right? Shall we follow this convention?
Coding convention of GNU is FYI : https://www.gnu.org/prep/standards/html_node/Names.html#Names  .

Comment (ZhijunZhao) [inline Lesson3/app/main.c]: I just created one task to follow the c naming conventions.

File: Lesson3/app/main.c [hunk 3]
```diff
@@ -75,9 +77,71 @@ static void sendMessageAndBlink(IOTHUB_CLIENT_LL_HANDLE iotHubClientHandle, char
...
+static char *readFile(char *fileName)
+{
+    FILE *fp;
+    long lSize;
+    char *buffer;
+
+    fp = fopen(fileName, "rb");
+    if (!fp)
+    {
+        printf("ERROR: %s doesn't exist!\r\n", fileName);
```
Comment (aovechki) [inline Lesson3/app/main.c]: >\r\n [](start = 40, length = 4)

some error messages include EOL, some do not. Why?

File: Lesson3/app/main.c [hunk 4]
```diff
@@ -75,9 +77,71 @@ static void sendMessageAndBlink(IOTHUB_CLIENT_LL_HANDLE iotHubClientHandle, char
...
+    sprintf(keyName, "%s/%s-key.pem", cwd, deviceId);
+
+    char *x509certificate = readFile(certName);
+    char *x509privatekey = readFile(keyName);
+
+    if (
+        (IoTHubClient_LL_SetOption(iotHubClientHandle, OPTION_X509_CERT, x509certificate) != IOTHUB_CLIENT_OK) ||
+        (IoTHubClient_LL_SetOption(iotHubClientHandle, OPTION_X509_PRIVATE_KEY, x509privatekey) != IOTHUB_CLIENT_OK))
+    {
+        printf("failure to set options for x509, aborting\r\n");
```
Comment (aovechki) [inline Lesson3/app/main.c]: >printf("failure to set options for x509, aborting\r\n"); [](start = 8, length = 56)

it's a bit inconsistent how we deal with console output. Sometimes messages are prefixed with "ERROR:", sometimes not. The same applies to the EOL. Should factor out and make them consistent.

File: Lesson3/app/main.c [hunk 5]
```diff
@@ -75,9 +77,71 @@ static void sendMessageAndBlink(IOTHUB_CLIENT_LL_HANDLE iotHubClientHandle, char
...
+    }
+
+    free(x509certificate);
+    free(x509privatekey);
```
Comment (aovechki) [inline Lesson3/app/main.c]: check for NULL first?

File: Lesson3/app/main.c (hunk 4 above)
Comment (aovechki) [inline Lesson3/app/main.c]: Also, i can't find where the code is 'aborting'


---
In reply to: [89926478](https://github.com/Azure-Samples/iot-hub-c-edison-getting-started/pull/20#discussion_r89926478) [](ancestors = 89926478)

File: Lesson3/app/main.c [hunk 6]
```diff
@@ -94,15 +158,21 @@ int main(int argc, char *argv[])
...
         IOTHUB_CLIENT_LL_HANDLE iotHubClientHandle;
-        if ((iotHubClientHandle = IoTHubClient_LL_CreateFromConnectionString(argv[2], MQTT_Protocol)) == NULL)
+        if ((iotHubClientHandle = IoTHubClient_LL_CreateFromConnectionString(argv[1], MQTT_Protocol)) == NULL)
         {
-            (void)printf("ERROR: iotHubClientHandle is NULL!\r\n");
+            printf("ERROR: iotHubClientHandle is NULL!\r\n");
         }
         else
         {
+            if (argc == 4 && strcmp(argv[3], "x509=true") == 0)
```
Comment (aovechki) [inline Lesson3/app/main.c]: >if (argc == 4 && strcmp(argv[3], "x509=true") == 0) [](start = 12, length = 51)

why not simply call strstr on the connection string?

File: Lesson3/app/main.c [hunk 7]
```diff
@@ -94,15 +158,21 @@ int main(int argc, char *argv[])
...
+            {
+                // Try using X.509 certificate.
```
Comment (aovechki) [inline Lesson3/app/main.c]: >Try using X.509 certificate. [](start = 19, length = 28)

=> 'Use X.509 certificate authentication'

**Comment by aovechki** [comment]
Are the files inside .vscode included by design?

File: Lesson3/app/main.c [hunk 8]
```diff
@@ -75,9 +77,71 @@ static void sendMessageAndBlink(IOTHUB_CLIENT_LL_HANDLE iotHubClientHandle, char
...
+    fseek(fp, 0L, SEEK_END);
+    lSize = ftell(fp);
+    rewind(fp);
+
+    /* allocate memory for entire content */
+    buffer = calloc(1, lSize + 1);
+    if (!buffer)
+    {
+        fclose(fp);
+        printf("ERROR: memory alloc fails");
```
Comment (aovechki) [inline Lesson3/app/main.c]: >emory alloc fails [](start = 24, length = 17)

=> Failed to allocate memory

File: Lesson3/app/main.c [hunk 9]
```diff
@@ -75,9 +77,71 @@ static void sendMessageAndBlink(IOTHUB_CLIENT_LL_HANDLE iotHubClientHandle, char
...
+        return NULL;
+    }
+
+    /* copy the file into the buffer */
+    if (1 != fread(buffer, lSize, 1, fp))
+    {
+        fclose(fp);
+        free(buffer);
+        printf("ERROR: entire read fails");
```
Comment (aovechki) [inline Lesson3/app/main.c]: >entire read fails [](start = 23, length = 17)

Failed to read the file %s into memory

File: Lesson3/app/main.c [hunk 10]
```diff
@@ -94,15 +158,21 @@ int main(int argc, char *argv[])
...
         {
+            if (argc == 4 && strcmp(argv[3], "x509=true") == 0)
+            {
+                // Try using X.509 certificate.
+                setX509Certificate(iotHubClientHandle, argv[2]);
+            }
+
             if (IoTHubClient_LL_SetOption(iotHubClientHandle, "TrustedCerts", certificates) != IOTHUB_CLIENT_OK)
             {
                 printf("failure to set option \"TrustedCerts\"\r\n");
```
Comment (aovechki) [inline Lesson3/app/main.c]: >failure to set option \"TrustedCerts\"\r\n [](start = 24, length = 42)

=> 'Failed to set TrustedCerts option'

**Comment by ZhijunZhao** [comment]
Will focus on naming conventions in a separate PR.

**Pull request merged**